In [11]:
%cd /content
!rm -rf 18_10123154_12523066_CDUTV
!git clone https://github.com/YoungHyyy/18_10123154_12523066_CDUTV
%cd 18_10123154_12523066_CDUTV

import sys, sklearn, numpy, pandas
sys.path.insert(0, "ai-models/src")
print("sklearn", sklearn.__version__, "| numpy", numpy.__version__, "| pandas", pandas.__version__)

/content
Cloning into '18_10123154_12523066_CDUTV'...
remote: Enumerating objects: 47, done.
remote: Counting objects: 100% (47/47), done.
remote: Compressing objects: 100% (39/39), done.
remote: Total 47 (delta 11), reused 42 (delta 6), pack-reused 0 (from 0)
Receiving objects: 100% (47/47), 1.74 MiB | 8.19 MiB/s, done.
Resolving deltas: 100% (11/11), done.
/content/18_10123154_12523066_CDUTV
sklearn 1.6.1 | numpy 2.1.3 | pandas 2.2.3


In [12]:
import warnings; warnings.filterwarnings("ignore")
from preprocess import load_data, split, ROOT

X, y = load_data()
X_train, X_test, y_train, y_test = split(X, y)
print("Train:", X_train.shape, "| Test:", X_test.shape)
print("Tỉ lệ ác tính - train:", round(y_train.mean(), 3), "| test:", round(y_test.mean(), 3))

Train: (455, 30) | Test: (114, 30)
Tỉ lệ ác tính - train: 0.374 | test: 0.368


In [13]:
from train import run_experiments

results, fitted = run_experiments(X_train, X_test, y_train, y_test)
results

>>> Logistic Regression (baseline)
>>> KNN
>>> SVM (RBF)
>>> Decision Tree
>>> Random Forest


,model,best_params,cv_f1_mean,cv_f1_std,train_recall,train_f1,test_recall,test_precision,test_f1,test_accuracy,test_roc_auc,test_pr_auc,fit_seconds,predict_ms_per_sample,size_kb
0,Logistic Regression (baseline),{'C': 0.1},0.9671,0.0112,0.9765,0.9794,0.9762,1.0000,0.9880,0.9912,0.9977,0.9966,2.4,0.029,2.4
1,KNN,"{'n_neighbors': 3, 'weights': 'uniform'}",0.9633,0.0250,0.9588,0.9790,0.8571,0.9730,0.9114,0.9386,0.9825,0.9712,0.8,0.036,102.3
2,SVM (RBF),"{'C': 10, 'gamma': 0.01}",0.9730,0.0116,0.9706,0.9851,0.9762,1.0000,0.9880,0.9912,0.9970,0.9958,1.9,0.028,16.6
3,Decision Tree,"{'criterion': 'gini', 'max_depth': 5, 'min_sam...",0.9378,0.0297,0.9706,0.9510,0.8333,0.8974,0.8642,0.9035,0.9076,0.9120,3.2,0.032,3.2
4,Random Forest,"{'max_depth': None, 'min_samples_leaf': 1, 'n_...",0.9525,0.0175,1.0000,1.0000,0.9286,1.0000,0.9630,0.9737,0.9965,0.9943,44.9,0.330,89.4


In [18]:
cols = ["model", "best_params", "cv_f1_mean", "cv_f1_std", "fit_seconds",
        "predict_ms_per_sample", "size_kb", "train_f1",
        "test_recall", "test_precision", "test_f1", "test_roc_auc"]
print(results[cols].to_csv(index=False))

model,best_params,cv_f1_mean,cv_f1_std,fit_seconds,predict_ms_per_sample,size_kb,train_f1,test_recall,test_precision,test_f1,test_roc_auc
Logistic Regression (baseline),{'C': 0.1},0.9671,0.0112,2.4,0.029,2.4,0.9794,0.9762,1.0,0.988,0.9977
KNN,"{'n_neighbors': 3, 'weights': 'uniform'}",0.9633,0.025,0.8,0.036,102.3,0.979,0.8571,0.973,0.9114,0.9825
SVM (RBF),"{'C': 10, 'gamma': 0.01}",0.973,0.0116,1.9,0.028,16.6,0.9851,0.9762,1.0,0.988,0.997
Decision Tree,"{'criterion': 'gini', 'max_depth': 5, 'min_samples_leaf': 5}",0.9378,0.0297,3.2,0.032,3.2,0.951,0.8333,0.8974,0.8642,0.9076
Random Forest,"{'max_depth': None, 'min_samples_leaf': 1, 'n_estimators': 100}",0.9525,0.0175,44.9,0.33,89.4,1.0,0.9286,1.0,0.963,0.9965



In [14]:
cols = ["model", "test_recall", "test_precision", "test_f1", "test_roc_auc",
        "train_f1", "predict_ms_per_sample", "size_kb"]
results[cols].sort_values("test_recall", ascending=False)

,model,test_recall,test_precision,test_f1,test_roc_auc,train_f1,predict_ms_per_sample,size_kb
0,Logistic Regression (baseline),0.9762,1.0000,0.9880,0.9977,0.9794,0.029,2.4
2,SVM (RBF),0.9762,1.0000,0.9880,0.9970,0.9851,0.028,16.6
4,Random Forest,0.9286,1.0000,0.9630,0.9965,1.0000,0.330,89.4
1,KNN,0.8571,0.9730,0.9114,0.9825,0.9790,0.036,102.3
3,Decision Tree,0.8333,0.8974,0.8642,0.9076,0.9510,0.032,3.2


In [15]:
results[["model", "best_params", "cv_f1_mean", "cv_f1_std", "fit_seconds"]]

,model,best_params,cv_f1_mean,cv_f1_std,fit_seconds
0,Logistic Regression (baseline),{'C': 0.1},0.9671,0.0112,2.4
1,KNN,"{'n_neighbors': 3, 'weights': 'uniform'}",0.9633,0.0250,0.8
2,SVM (RBF),"{'C': 10, 'gamma': 0.01}",0.9730,0.0116,1.9
3,Decision Tree,"{'criterion': 'gini', 'max_depth': 5, 'min_sam...",0.9378,0.0297,3.2
4,Random Forest,"{'max_depth': None, 'min_samples_leaf': 1, 'n_...",0.9525,0.0175,44.9


In [16]:
print("Nếu luôn đoán 'lành tính': Accuracy =", round(1 - y_test.mean(), 3), "| Recall ác tính = 0")

Nếu luôn đoán 'lành tính': Accuracy = 0.632 | Recall ác tính = 0


In [17]:
from google.colab import files

results.to_csv(ROOT / "docs" / "model_comparison.csv", index=False)
files.download("docs/model_comparison.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Huấn luyện và tinh chỉnh 5 model

**Cách làm chung (để so sánh công bằng):**
- Cùng bộ dữ liệu, cùng cách chia: 80% train / 20% test (455 / 114 mẫu), `stratify=y`, `random_state=42`.
- Mỗi model nằm trong một `Pipeline` gồm imputer, StandardScaler và model, nên các bước tiền xử lý chỉ được fit trên tập train (không rò rỉ dữ liệu).
- Tinh chỉnh siêu tham số bằng `GridSearchCV` với 5-fold cross-validation **chỉ trên tập train**, tối ưu theo F1 của lớp ác tính.
- Tập test chỉ dùng **một lần** để lấy kết quả cuối.
- Metric trọng tâm: **Recall của lớp ác tính**, vì bỏ sót ca ung thư (FN) nguy hiểm hơn báo nhầm (FP). Ngoài ra dùng Precision, F1, ROC-AUC.
- Ghi chú: "thời gian huấn luyện" bên dưới là tổng thời gian chạy toàn bộ `GridSearchCV` (gồm mọi tổ hợp tham số × 5 fold), không phải một lần fit. "Thời gian dự đoán" là trung bình trên mỗi mẫu của tập test.

### 1. Logistic Regression (baseline)
- **Nguyên lý:** tính tổ hợp tuyến tính của 30 đặc trưng rồi đưa qua hàm sigmoid để ra xác suất ác tính; xác suất ≥ 0.5 thì dự đoán ác tính. Ranh giới quyết định là một siêu phẳng.
- **Vì sao làm baseline:** đơn giản, nhanh, dễ giải thích; là mốc để biết các model phức tạp hơn có thực sự đáng giá không.
- **Siêu tham số:** `C` (nghịch đảo độ mạnh của điều chuẩn L2). Đã thử C ∈ {0.01, 0.1, 1, 10, 100}. **Chọn C = 0.1**, tức điều chuẩn khá mạnh. C nhỏ thì model đơn giản, dễ underfit; C lớn thì điều chuẩn yếu, dễ overfit. `class_weight="balanced"` để bù lệch lớp nhẹ (khoảng 63% lành / 37% ác).
- **Thời gian huấn luyện (cả GridSearch):** 2.4 giây | **Dự đoán:** 0.029 ms/mẫu | **Kích thước file:** 2.4 KB.
- **Cross-validation:** F1 trung bình 0.9671, độ lệch chuẩn 0.0112.

### 2. KNN
- **Nguyên lý:** không học tham số. Với mẫu mới, tìm K mẫu train gần nhất (khoảng cách Euclid trên dữ liệu đã chuẩn hóa) rồi bỏ phiếu theo đa số.
- **Vì sao cần chuẩn hóa:** khoảng cách bị chi phối bởi cột có thang đo lớn (ví dụ `area` hàng trăm so với `smoothness` dưới 1).
- **Siêu tham số:** `n_neighbors` ∈ {3, 5, 7, 9, 11, 15}, `weights` ∈ {uniform, distance}. **Chọn K = 3, weights = uniform.** K nhỏ thì biên phức tạp, nhạy nhiễu, dễ overfit; K lớn thì biên mượt, dễ underfit. K = 3 là giá trị nhỏ nhất trong lưới, cho thấy model bám sát dữ liệu train và có xu hướng overfit (Train F1 0.979 so với Test F1 0.911).
- **Thời gian huấn luyện (cả GridSearch):** 0.8 giây | **Dự đoán:** 0.036 ms/mẫu | **Kích thước file:** 102.3 KB (lớn nhất, vì lưu cả dữ liệu train).
- **Cross-validation:** F1 trung bình 0.9633, độ lệch chuẩn 0.0250 (kém ổn định hơn Logistic và SVM).

### 3. SVM (kernel RBF)
- **Nguyên lý:** tìm ranh giới có **lề lớn nhất** giữa hai lớp. Kernel RBF ánh xạ ngầm dữ liệu sang không gian nhiều chiều để tách được ranh giới phi tuyến.
- **Siêu tham số:** `C` ∈ {0.1, 1, 10, 100} (đánh đổi giữa lề rộng và số điểm bị phân loại sai), `gamma` ∈ {scale, 0.01, 0.1} (độ "cong" của biên; gamma lớn thì mỗi điểm ảnh hưởng hẹp, biên uốn lượn, dễ overfit). **Chọn C = 10, gamma = 0.01**: gamma nhỏ cho biên mượt, C = 10 cho phép một số sai số để lề đủ rộng.
- `probability=True` để có xác suất phục vụ ROC-AUC và hiển thị trên app; `class_weight="balanced"`.
- **Thời gian huấn luyện (cả GridSearch):** 1.9 giây | **Dự đoán:** 0.028 ms/mẫu | **Kích thước file:** 16.6 KB.
- **Cross-validation:** F1 trung bình 0.9730 (cao nhất trong 5 model), độ lệch chuẩn 0.0116.

### 4. Decision Tree
- **Nguyên lý:** chia dữ liệu liên tiếp bằng các điều kiện dạng "đặc trưng ≤ ngưỡng?". Mỗi lần chia chọn điều kiện làm giảm độ tạp chất (Gini hoặc Entropy) nhiều nhất.
- **Ưu điểm:** dễ giải thích (có thể vẽ thành cây), không cần chuẩn hóa. **Nhược điểm:** cây sâu rất dễ overfit.
- **Siêu tham số:** `criterion` ∈ {gini, entropy}, `max_depth` ∈ {3, 5, 7, 10, None}, `min_samples_leaf` ∈ {1, 2, 5}. Giới hạn độ sâu và số mẫu tối thiểu ở lá chính là cách hạn chế độ phức tạp của cây (pruning). **Chọn criterion = gini, max_depth = 5, min_samples_leaf = 5.**
- **Thời gian huấn luyện (cả GridSearch):** 3.2 giây (lưới có 30 tổ hợp) | **Dự đoán:** 0.032 ms/mẫu | **Kích thước file:** 3.2 KB.
- **Cross-validation:** F1 trung bình 0.9378 (thấp nhất), độ lệch chuẩn 0.0297 (kém ổn định nhất).

### 5. Random Forest
- **Nguyên lý:** huấn luyện nhiều cây, mỗi cây học trên một mẫu bootstrap của dữ liệu và chỉ xét ngẫu nhiên một tập đặc trưng ở mỗi lần chia; kết quả bỏ phiếu theo đa số (bagging). Trung bình nhiều cây làm giảm phương sai so với một cây đơn.
- **Siêu tham số:** `n_estimators` ∈ {100, 200, 300}, `max_depth` ∈ {None, 5, 10}, `min_samples_leaf` ∈ {1, 2}. **Chọn n_estimators = 100, max_depth = None, min_samples_leaf = 1.** Rừng đã ổn định với 100 cây nên thêm cây không cải thiện thêm.
- **Thời gian huấn luyện (cả GridSearch):** 44.9 giây (lâu nhất, vì 18 tổ hợp × 5 fold, mỗi lần fit 100 đến 300 cây) | **Dự đoán:** 0.33 ms/mẫu (chậm nhất) | **Kích thước file:** 89.4 KB.
- **Cross-validation:** F1 trung bình 0.9525, độ lệch chuẩn 0.0175.

### Nhận xét sơ bộ từ bảng kết quả

| Model | Train F1 | Test F1 | Chênh (train − test) | Test Recall | Test Precision |
|---|---|---|---|---|---|
| Logistic Regression (baseline) | 0.9794 | 0.9880 | −0.009 | 0.9762 | 1.0000 |
| KNN | 0.9790 | 0.9114 | 0.068 | 0.8571 | 0.9730 |
| SVM (RBF) | 0.9851 | 0.9880 | −0.003 | 0.9762 | 1.0000 |
| Decision Tree | 0.9510 | 0.8642 | 0.087 | 0.8333 | 0.8974 |
| Random Forest | 1.0000 | 0.9630 | 0.037 | 0.9286 | 1.0000 |

- **Recall test cao nhất:** Logistic Regression và SVM (cùng 0.9762).
- **Ổn định nhất (`cv_f1_std` nhỏ nhất):** Logistic Regression (0.0112), sát SVM (0.0116).
- **Có dấu hiệu overfit:** Decision Tree (chênh 0.087), KNN (chênh 0.068), Random Forest (Train F1 = 1.0 nhưng Test F1 = 0.963). Logistic Regression và SVM có Test F1 không thấp hơn Train F1, tức không overfit. Việc Test F1 nhỉnh hơn Train F1 một chút là do tập test nhỏ (114 mẫu), không phải dấu hiệu bất thường.
- **So với baseline Logistic Regression (Test F1 = 0.988):** không model nào vượt baseline trên tập test; SVM chỉ ngang bằng. Điều này cho thấy ranh giới giữa hai lớp gần như tuyến tính, khớp với pairplot ở phần EDA (hai lớp tách khá rõ), nên model đơn giản đã đủ tốt.